sdk/core/azure-core/src/http/telemetry_policy.cpp

```diff
@@ -13,6 +13,10 @@ std::unique_ptr<RawResponse> Azure::Core::Http::Policies::_internal::TelemetryPo
     NextHttpPolicy nextPolicy,
     Context const& context) const
 {
-  request.SetHeader("User-Agent", m_telemetryId);
```

Comment:
I would argue that having a separate policy which sets User-Agent is a coding error. The Azure SDK has fairly strict requirements on user-agent, if the telemetry_policy is in the pipeline, it means that developers expect that the pipeline will implement the Azure SDK guidelines.

If we have tests which depend on the user-agent policy not being in the pipeline, we should change the tests.

Reply:
This is about bringing codegen up to spec.

Autorest.testserver has a test (https://github.com/Azure/autorest.testserver/blob/main/swagger/header.json#L13-L31) that specifically check that if the swagger says that `User-Agent` should have some specific value, the generated code will send that and nothing else. Not even appending. Java, .NET, and Python do pass that test. They have logic to intentionally enable this scenario.

There is another similar test, for `x-ms-client-request-id`, C++ fails it, but Go, .NET, and Python do pass.

The test specifies these headers intentionally, there is no other way to read it. For example, for User-Agent, the value should be `"overwrite"`, i.e. "overwrite the user-agent". As almost it expects that naive policy implementation would simply overwrite the values, and wants to prevent that.

Reply:
Why can't the test configure the pipeline with the desired set of policies, excluding the TelemetryPolicy and RequestActivityPolicy?

Reply:
It is not the SDK test, it is codegen test. Codegen generates client with the default set of options. It should not be generating new pipeline without a policy just for one method that happens to use `User-Agent` or `x-ms-client-request-id`.

Reply:
For example, for that user-agent test, the following code gets generated:

```
Response<ParamHeaderExistingKeyResult> HeaderClient::ParamExistingKey(
      const ParamHeaderExistingKeyOptions& options,
      const Core::Context& context)
  {
    auto url = m_url;
    url.AppendPath("header/param/existingkey");
    auto request = Core::Http::Request(Core::Http::HttpMethod::Post, url);
    if (!options.UserAgent.empty())
    {
      request.SetHeader("User-Agent", options.UserAgent);
    }
    auto pRawResponse = m_pipeline->Send(request, context);
    auto httpStatusCode = pRawResponse->GetStatusCode();
    if (httpStatusCode != Core::Http::HttpStatusCode::Ok)
    {
      throw Core::RequestFailedException(pRawResponse);
    }
    ParamHeaderExistingKeyResult response;
    return Response<ParamHeaderExistingKeyResult>(std::move(response), std::move(pRawResponse));
  }
```

Reply:
In the sample above, where did m_pipeline come from? Presumably it was created by test code, which we control.

Btw, I checked the .Net implementation for both TelemetryPolicy and RequestActivityPolicy - TelemetryPolicy has a similar "if not set use the default" behavior you describe. RequestActivityPolicy is somewhat different - it depends on the ClientRequestId property in the Request object, which is defaulted in the constructor to be a Guid.NewGuid.

Reply:
It is not created by the test code, it is created by codegen for clients. Codegen does not know whether it generates client for tests or for an Azure service.
It gets generated as such:

Header -
```cpp
  class HeaderClient final {
  private:
    std::shared_ptr<Azure::Core::Http::_internal::HttpPipeline> m_pipeline;
    Core::Url m_url;

  public:
    HeaderClient();

    Response<ParamHeaderExistingKeyResult> ParamExistingKey(
        const ParamHeaderExistingKeyOptions& options,
        const Core::Context& context = {});

    Response<ResponseHeaderExistingKeyResult> ResponseExistingKey(
        const Core::Context& context = {});

    Response<ParamHeaderProtectedKeyResult> ParamProtectedKey(
        const ParamHeaderProtectedKeyOptions& options,
        const Core::Context& context = {});

    Response<ResponseHeaderProtectedKeyResult> ResponseProtectedKey(
        const Core::Context& context = {});

    Response<ParamHeaderIntegerResult> ParamInteger(
        const ParamHeaderIntegerOptions& options,
        const Core::Context& context = {});

    Response<ResponseHeaderIntegerResult> ResponseInteger(
        const ResponseHeaderIntegerOptions& options,
        const Core::Context& context = {});

    Response<ParamHeaderLongResult> ParamLong(
        const ParamHeaderLongOptions& options,
        const Core::Context& context = {});

    Response<ResponseHeaderLongResult> ResponseLong(
        const ResponseHeaderLongOptions& options,
        const Core::Context& context = {});

    Response<ParamHeaderFloatResult> ParamFloat(
        const ParamHeaderFloatOptions& options,
        const Core::Context& context = {});

    Response<ResponseHeaderFloatResult> ResponseFloat(
        const ResponseHeaderFloatOptions& options,
        const Core::Context& context = {});

    Response<ParamHeaderDoubleResult> ParamDouble(
        const ParamHeaderDoubleOptions& options,
        const Core::Context& context = {});

    Response<ResponseHeaderDoubleResult> ResponseDouble(
        const ResponseHeaderDoubleOptions& options,
        const Core::Context& context = {});

    Response<ParamHeaderBoolResult> ParamBool(
        const ParamHeaderBoolOptions& options,
        const Core::Context& context = {});

    Response<ResponseHeaderBoolResult> ResponseBool(
        const ResponseHeaderBoolOptions& options,
        const Core::Context& context = {});

    Response<ParamHeaderStringResult> ParamString(
        const ParamHeaderStringOptions& options,
        const Core::Context& context = {});

    Response<ResponseHeaderStringResult> ResponseString(
        const ResponseHeaderStringOptions& options,
        const Core::Context& context = {});

    Response<ParamHeaderDateResult> ParamDate(
        const ParamHeaderDateOptions& options,
        const Core::Context& context = {});

    Response<ResponseHeaderDateResult> ResponseDate(
        const ResponseHeaderDateOptions& options,
        const Core::Context& context = {});

    Response<ParamHeaderDatetimeResult> ParamDatetime(
        const ParamHeaderDatetimeOptions& options,
        const Core::Context& context = {});

    Response<ResponseHeaderDatetimeResult> ResponseDatetime(
        const ResponseHeaderDatetimeOptions& options,
        const Core::Context& context = {});

    Response<ParamHeaderDatetimeRfc1123Result> ParamDatetimeRfc1123(
        const ParamHeaderDatetimeRfc1123Options& options,
        const Core::Context& context = {});

    Response<ResponseHeaderDatetimeRfc1123Result> ResponseDatetimeRfc1123(
        const ResponseHeaderDatetimeRfc1123Options& options,
        const Core::Context& context = {});

    Response<ParamHeaderDurationResult> ParamDuration(
        const ParamHeaderDurationOptions& options,
        const Core::Context& context = {});

    Response<ResponseHeaderDurationResult> ResponseDuration(
        const ResponseHeaderDurationOptions& options,
        const Core::Context& context = {});

    Response<ParamHeaderByteResult> ParamByte(
        const ParamHeaderByteOptions& options,
        const Core::Context& context = {});

    Response<ResponseHeaderByteResult> ResponseByte(
        const ResponseHeaderByteOptions& options,
        const Core::Context& context = {});

    Response<ParamHeaderEnumResult> ParamEnum(
        const ParamHeaderEnumOptions& options,
        const Core::Context& context = {});

    Response<ResponseHeaderEnumResult> ResponseEnum(
        const ResponseHeaderEnumOptions& options,
        const Core::Context& context = {});

    Response<CustomHeaderRequestIdResult> CustomRequestId(const Core::Context& context = {});
  };
```

.cpp file -
```cpp
  HeaderClient::HeaderClient()
      : m_pipeline(
          new Core::Http::_internal::HttpPipeline({}, "azure-test-autorest-header", "0", {}, {})),
        m_url("http://localhost:3000")
  {
  }
```

Reply:
In my response above, I thought I described the behavior for the .Net SDK: It initializes the ClientRequestId to a GUID in the Request constructor,  and then blindly overwrites any existing x-ms-client-request-id header. That's why you see the "preservation" behavior (which isn't *quite* preservation since it will overwrite any headers set in the headers map).  And the TelemetryPolicy uses the UserAgent header set by the client if it's set.

FWIW, I usually try to track down the implementation of the functionality rather than the tests because often times the tests have hidden assumptions that may not be backed up in live code (for instance dealing with mocked HTTP pipelines or transports).

Reply:
Here is `x-ms-client-request-id` preservation logic in the Azure SDK for Java:
https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/core/azure-core/src/main/java/com/azure/core/http/policy/RequestIdPolicy.java#L40-L43